src/tree/identity_tree.rs

```
@@ -0,0 +1,592 @@
+
+        if let Some(root) = root {
+            if root.hash == self.tree.root() {
+                let proof = self.tree.proof(*leaf_idx as usize);
+                Ok(Some(InclusionProof::new(self.tree.root(), proof)))
+            } else {
+                let proof = self.construct_proof_from_root(*leaf_idx, root)?;
+                Ok(Some(InclusionProof::new(root.hash, proof)))
+            }
+        } else {
+            let proof = self.tree.proof(*leaf_idx as usize);
+            Ok(Some(InclusionProof::new(self.tree.root(), proof)))
+        }
+    }
+
+    pub fn construct_proof_from_root(
+        &self,
+        leaf_idx: u32,
+        root: &Root,
+    ) -> eyre::Result<Proof> {
+        let updates = self
+            .tree_updates
+            .get(root)
+            .ok_or_eyre("Could not find root in tree updates")?;
+
+        let mut node_idx = leaf_to_storage_idx(leaf_idx, self.tree.depth());
+
+        let mut proof: Vec<Branch<Hash>> = vec![];
+
+        while node_idx > 0 {
+            let sibling_idx = if node_idx % 2 == 0 {
+                node_idx - 1
+            } else {
+                node_idx + 1
+            };
+
+            let sibling = updates
+                .get(&sibling_idx)
+                .copied()
+                .or_else(|| {
+                    let (depth, offset) =
+                        storage_idx_to_coords(sibling_idx as usize);
+                    Some(self.tree.get_node(depth, offset))
+                })
+                .expect("Could not find node in tree");
+
+            proof.push(if node_idx % 2 == 0 {
+                Branch::Right(sibling)
+            } else {
+                Branch::Left(sibling)
+            });
+
+            node_idx = (node_idx - 1) / 2;
+        }
+
+        Ok(semaphore::merkle_tree::Proof(proof))
+    }
+
+    pub fn insert(&mut self, index: u32, value: Hash) {
+        //TODO: do we want to return an error if a non unique leaf is being inserted?
```

Is there any reason why we wouldn't want to return an error in that case?